hipamd/src/hip_mrfs.cpp

```diff
@@ -24,208 +28,610 @@
 struct MemcpyTask {
   void* dst;
   const void* src;
   size_t sizeBytes;
   hipMemcpyKind kind;
   hip::Stream* stream;
-  int deviceId;
+  QueueId queueId;
   std::atomic<bool> completed{false};
 };
 
-// Thread pool for paralell memory operations
+// quota management
+class QuotaManager {
+  private:
+  // Store bandwidth quotas in bytes per second for each (device, process) queue
+  std::map<QueueId, double> _queueQuotas;
+  std::map<QueueId, double> _queueUsage;
+  std::mutex _mutex;
+  std::thread _resetThread;
+  std::atomic<bool> _stopResetThread{false};
+  double _totalProcessQuota;
+  std::map<pid_t, double> _processQuotas;
+  std::map<pid_t, double> _processUsage;  // Track usage at process level
+
+  void quotaResetThreadFunction() {
+    MRFS_LOG("QUOTA", "Quota reset thread started");
+    while (!_stopResetThread) {
+      // Sleep for 1 second
+      struct timespec ts;
+      ts.tv_sec = 1;  // 1 second
+      ts.tv_nsec = 0;
+
+      nanosleep(&ts, NULL);
+
+      // Reset all queue usage
+      std::lock_guard<std::mutex> lock(_mutex);
+      for (auto& usage : _queueUsage) {
+        double oldUsage = usage.second;
+        usage.second = 0.0;
+        MRFS_LOG("QUOTA",
+                 "Reset bandwidth for queue (device=" << usage.first.deviceId
+                                                      << ", process=" << usage.first.processId
+                                                      << ") from " << oldUsage << " B/s to 0 B/s");
+      }
+
+      // Reset process usage too
+      for (auto& usage : _processUsage) {
+        double oldUsage = usage.second;
+        usage.second = 0.0;
+        MRFS_LOG("QUOTA",
+                 "Reset bandwidth for process " << usage.first << " from " << oldUsage
+                                                << " B/s to 0 B/s");
+      }
+    }
+  }
+
+  QuotaManager() : _totalProcessQuota(DEFAULT_PROCESS_BANDWIDTH_QUOTA) {
+    // Start the reset thread
+    _resetThread = std::thread(&QuotaManager::quotaResetThreadFunction, this);
+  }
+
+  // Destructor to clean up thread
+  ~QuotaManager() {
+    _stopResetThread = true;
+    if (_resetThread.joinable()) {
+      _resetThread.join();
+    }
+  }
+
+  public:
+  static QuotaManager& instance() {
+    static QuotaManager manager;
+    return manager;
+  }
+
+  // Set total quota for a process (in bytes per second)
+  void setProcessQuota(pid_t processId, double bandwidthBytesPerSec) {
+    std::lock_guard<std::mutex> lock(_mutex);
+    _processQuotas[processId] = bandwidthBytesPerSec;
+    _processUsage[processId] = 0.0;  // Initialize process usage
+
+    // Reallocate quotas for all queues associated with this process
+    reallocateQuotas(processId);
+
+    MRFS_LOG("QUOTA",
+             "Set total bandwidth quota for process " << processId << " to " << bandwidthBytesPerSec
+                                                      << " B/s");
+  }
+
+  void reallocateQuotas(pid_t processId) {
```

Comment:
TODO: add notion of "unused" devices. Need to explore other methods of tracking devices being used

Reply:
Instead of checking number of queues, check for how many queues with size >0